### PR TITLE
Fix checkpointed state for lr_schedulers with step interval

### LINKED ISSUE #7637: LR scheduler steps after saving checkpoint with iteration-based checkpointing
## 🐛 Bug

When checkpoints are saved using every_n_train_steps argument of ModelCheckpoint callback, checkpointing happens before lr_scheduler step being called. While it is the other way around with every_n_val_epochs argument.

In training_loop, on_train_batch_end (which saves checkpoint) is called before updating lr_scheduler. https://github.com/PyTorchLightning/pytorch-lightning/blob/master/pytorch_lightning/trainer/training_loop.py#L500-L529

while on_evaluation_end is called after updating lrs. https://github.com/PyTorchLightning/pytorch-lightning/blob/master/pytorch_lightning/trainer/trainer.py#L996-L1011

## Please reproduce using the BoringModel

https://colab.research.google.com/drive/1bBkhGiKJoavp1O4Oi4kLWVnxohl5HR6M?usp=sharing

### Expected behavior

In the colab notebook, a training epoch is of 64 batches. When I save checkpoint using every_n_train_steps=64 or with every_n_val_epochs=1, the lr_scheduler step count should be the same. But the step count is 1 lesser with every_n_train_steps=64

### Environment

 - PyTorch Version (e.g., 1.0):
 - OS (e.g., Linux):
 - How you installed PyTorch (`conda`, `pip`, source):
 - Build command you used (if compiling from source):
 - Python version:
 - CUDA/cuDNN version:
 - GPU models and configuration:
 - Any other relevant information:

### Additional context

cc @ananthsub

### PR DESCRIPTION
**Pull request by simran2905**
## What does this PR do?

Fixes #7637 

The lr schedulers are being updated after calling on_train_batch_end and trainer._run_evaluation(), both of which can save checkpoints. This means that the latest state of lr_scheduler is not saved during checkpoint. This PR fixes the issue for classic schedulers. 

NOTE: The ReduceLROnPlateau schedulers need on_train_batch_end called first for any metrics logging and will continue to save an older state potentially.

## Before submitting
- [X] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [X] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [X] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [X] Did you make sure to update the documentation with your changes? (if necessary)
- [X] Did you write any new necessary tests? (not for typos and docs)
- [X] Did you verify new and existing tests pass locally with your changes?
- [x] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [X] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [X] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: tests/trainer/optimization/test_optimizers.py [hunk 1]
```diff
@@ -620,3 +621,38 @@ def test_lr_scheduler_epoch_step_frequency(mocked_sched, check_val_every_n_epoch
     )
     trainer.fit(model)
     assert mocked_sched.call_count == expected_steps
+
+
+@pytest.mark.parametrize("every_n_train_steps", [None, 2])
+def test_scheduler_lr_step_interval_updated_before_saving(tmpdir, every_n_train_steps):
```
Comment (tchaton) [inline tests/trainer/optimization/test_optimizers.py:627]: Mind adding a test for `reduce_on_plateau` ?

**Review by tchaton — COMMENTED** [review]
Overall, looks good to me

File: pytorch_lightning/trainer/training_loop.py [hunk 2]
```diff
@@ -504,6 +504,9 @@ def run_training_epoch(self):
             if batch_output.signal == -1:
                 break
 
+            # update non-plateau LR schedulers
+            self.update_lr_schedulers('step', skip_plateau_scheduler=True)
```
Comment (carmocca) [inline pytorch_lightning/trainer/training_loop.py:508]: Why here?

Should be right above `VALIDATE IF NEEDED` (L526)

Comment (simran2905) [reply pytorch_lightning/trainer/training_loop.py:508]: Because every_n_train_steps based checkpointing happens in on_train_batch_end, so we need this before on_train_batch_end.
Specifically, in the test added every_n_train_steps=2 case will fail if this was added after on_train_batch_end.

**Comment by simran2905** [comment]
@carmocca Let me know if this looks good

File: pytorch_lightning/trainer/connectors/optimizer_connector.py [hunk 3]
```diff
@@ -46,6 +51,10 @@ def update_learning_rates(self, interval: str, opt_indices: Optional[List[int]]
             if isinstance(lr_scheduler['opt_idx'], int) and lr_scheduler['opt_idx'] not in opt_indices:
                 continue
 
+            if interval == "step" and ((skip_plateau_scheduler and lr_scheduler["reduce_on_plateau"]) or
+                                       (not skip_plateau_scheduler and not lr_scheduler["reduce_on_plateau"])):
```
Comment (carmocca) [inline pytorch_lightning/trainer/connectors/optimizer_connector.py:55]: Could use XOR

```suggestion
            if interval == "step" and not skip_plateau_scheduler ^ lr_scheduler["reduce_on_plateau"]:
```

File: pytorch_lightning/trainer/connectors/optimizer_connector.py [hunk 4]
```diff
@@ -29,12 +29,17 @@ def on_trainer_init(self) -> None:
...
-    def update_learning_rates(self, interval: str, opt_indices: Optional[List[int]] = None) -> None:
+    def update_learning_rates(
+        self, interval: str, opt_indices: Optional[List[int]] = None, skip_plateau_scheduler: bool = False
+    ) -> None:
         """Update learning rates.
 
         Args:
             interval: either 'epoch' or 'step'.
             opt_indices: indices of the optimizers to update.
+            skip_plateau_scheduler: boolean to control whether ReduceLROnPlateau scheduler is being updated.
```
Comment (carmocca) [inline pytorch_lightning/trainer/connectors/optimizer_connector.py:40]:
```suggestion
            skip_plateau_scheduler: control whether any ``ReduceLROnPlateau`` schedulers get updated.
```

File: pytorch_lightning/trainer/connectors/optimizer_connector.py [hunk 5]
```diff
@@ -46,6 +51,10 @@ def update_learning_rates(self, interval: str, opt_indices: Optional[List[int]]
             if isinstance(lr_scheduler['opt_idx'], int) and lr_scheduler['opt_idx'] not in opt_indices:
                 continue
 
+            if interval == "step" and ((skip_plateau_scheduler and lr_scheduler["reduce_on_plateau"]) or
```
Comment (carmocca) [inline pytorch_lightning/trainer/connectors/optimizer_connector.py:54]: Do we need to check that `interval == 'step'` here?

File: pytorch_lightning/trainer/training_loop.py [hunk 6]
```diff
@@ -837,15 +840,16 @@ def backward(self, result, optimizer, opt_idx, *args, **kwargs):
                 self.trainer.lightning_module._current_fx_name = "on_after_backward"
                 self.trainer.lightning_module.log_grad_norm(grad_norm_dict)
 
-    def update_lr_schedulers(self, interval: str) -> None:
-        if interval == "step":
+    def update_lr_schedulers(self, interval: str, skip_plateau_scheduler: bool = False) -> None:
+        if interval == "step" and skip_plateau_scheduler:
```
Comment (carmocca) [inline pytorch_lightning/trainer/training_loop.py:844]: Why add the extra condition?

File: pytorch_lightning/trainer/training_loop.py [hunk 7]
```diff
@@ -527,7 +530,7 @@ def run_training_epoch(self):
             # -----------------------------------------
             self.save_loggers_on_train_batch_end()
 
-            # update LR schedulers
+            # update plateau LR scheduler after metrics are logged
             self.update_lr_schedulers('step')
```
Comment (carmocca) [inline pytorch_lightning/trainer/training_loop.py:534]: Probably should add one more test (or update the existing ones) so both scheduler types are available to test for any duplicate `step` calls.

**Comment by carmocca** [comment]
Great job!

Also remember to update the `CHANGELOG.md` file

File: pytorch_lightning/trainer/training_loop.py (hunk 7 above)
Comment (awaelchli) [reply pytorch_lightning/trainer/training_loop.py:498]: say you have two schedulers, a regular one and a plateau. 

```
# here it updates the regular one, and skips the plateau
self.update_lr_schedulers('step', skip_plateau_scheduler=True)
...
# here how do you know you already stepped the first scheduler?
self.update_lr_schedulers('step')
```

Comment (carmocca) [reply pytorch_lightning/trainer/training_loop.py:498]: Because the current implementation is using a XOR for `skip_plateau` and `is_plateau`.

So the first call sets `skip_plateau=True` so only when `not is_plateau` will get updated

And the second call has `skip_plateau=False` so only when `is_plateau` gets updated

Comment (simran2905) [reply pytorch_lightning/trainer/training_loop.py:498]: Maybe skip_plateau_scheduler is not the best naming. It seems we are not skipping (updating everything) when skip is False. I'll rename is update_plateau_schedulers.

File: pytorch_lightning/trainer/training_loop.py [hunk 8]
```diff
@@ -463,6 +463,12 @@ def run_training_epoch(self):
             if batch_output.signal == -1:
                 break
 
+            # update non-plateau LR schedulers
+            # update epoch-interval ones only when we are at the end of training epoch
+            self.update_lr_schedulers('step', update_plateau_schedulers=False)
+            if self._num_training_batches_reached(is_last_batch):
+                self.update_lr_schedulers('epoch', update_plateau_schedulers=False)
```
Comment (simran2905) [inline pytorch_lightning/trainer/training_loop.py:470]: @carmocca To make sure epoch based lr_schedulers are updated before validation (and only once at the end of training epoch), I added this condition. Would this cover all cases or does it need to be refined more?

**Comment by simran2905** [comment]
It seems tests passing locally are failing on CI without any local changes. What could be the reason?
Screen Shot 2021-06-15 at 1 25 45 PM

**Comment by awaelchli** [comment]
@simran2905 we updated the training loop today. The file you are editing is not anymore included in the tests. Let me update your branch by rebasing the changes so you can continue the work!

**Comment by awaelchli** [comment]
moved your changes from `pytorch_lightning/trainer/training_loop.py` to `pytorch_lightning/loops/training_epoch_loop.py`

File: pytorch_lightning/trainer/connectors/optimizer_connector.py [hunk 9]
```diff
@@ -29,12 +29,17 @@ def on_trainer_init(self) -> None:
...
+        self, interval: str, opt_indices: Optional[List[int]] = None, update_plateau_schedulers: bool = False
+    ) -> None:
         """Update learning rates.
 
         Args:
             interval: either 'epoch' or 'step'.
             opt_indices: indices of the optimizers to update.
+            update_plateau_schedulers: control whether ``ReduceLROnPlateau`` or non-plateau schedulers get updated.
+                This is required to make sure correct scheduler state is saved during checkpoint.
+                Non-plateau schedulers get updated before saving checkpoints.
```
Comment (carmocca) [inline pytorch_lightning/trainer/connectors/optimizer_connector.py:42]:
```suggestion
                Non-plateau schedulers get updated before running validation
```

File: tests/checkpointing/test_model_checkpoint.py [hunk 10]
```diff
@@ -262,6 +261,13 @@ def _make_assertions(epoch, ix, version=''):
         global_ix = ix + per_epoch_val_checks * epoch
         duplicated = bool(version)
 
+        # checkpoint saved at the end of training epoch will have updated
+        # lr_scheduler states
```
Comment (carmocca) [inline tests/checkpointing/test_model_checkpoint.py:265]: nit:

```suggestion
        # checkpoint saved at the end of training epoch will have updated lr_scheduler states
```

File: tests/trainer/optimization/test_optimizers.py [hunk 11]
```diff
@@ -620,3 +621,97 @@ def test_lr_scheduler_epoch_step_frequency(mocked_sched, check_val_every_n_epoch
...
+
+        def on_save_checkpoint(self, checkpoint):
+            lr_dict = checkpoint['lr_schedulers'][0]
+            # 2 batches ran. since the lr_dict interval is `step`, the step count should be 2
+            assert self.trainer.global_step + 1 == batches  # the global step hasn't been increased yet
+
+            if epoch_interval:
+                compare_to = max_epochs
+            else:
+                compare_to = batches
```
Comment (carmocca) [inline tests/trainer/optimization/test_optimizers.py:659]: nit: one line

```suggestion
            compare_to = max_epochs if epoch_interval else batches            
```

File: tests/trainer/optimization/test_optimizers.py [hunk 12]
```diff
@@ -620,3 +621,97 @@ def test_lr_scheduler_epoch_step_frequency(mocked_sched, check_val_every_n_epoch
...
+
+            lr_scheduler2 = torch.optim.lr_scheduler.StepLR(optimizer2, step_size=1)
+            lr_dict_2 = {'scheduler': lr_scheduler2, 'interval': 'step'}
+            return ({
+                'optimizer': optimizer1,
+                'lr_scheduler': lr_dict_1
+            }, {
+                'optimizer': optimizer2,
+                'lr_scheduler': lr_dict_2
+            })
```
Comment (carmocca) [inline tests/trainer/optimization/test_optimizers.py:702]: nit: simpler return

```suggestion
            return [optimizer1, optimizer2], [lr_dict_1, lr_dict_2]
```

File: pytorch_lightning/loops/training_epoch_loop.py [hunk 13]
```diff
@@ -347,7 +353,7 @@ def _prepare_outputs(
             processed_outputs = processed_outputs[0]
         return processed_outputs
 
-    def update_lr_schedulers(self, interval: str) -> None:
+    def update_lr_schedulers(self, interval: str, update_plateau_schedulers: bool = False) -> None:
```
Comment (carmocca) [inline pytorch_lightning/loops/training_epoch_loop.py:356]: The new parameter should be either `Optional[bool]` to update both or a required argument. Leaning towards the second option

File: pytorch_lightning/trainer/training_loop.py (hunk 8 above)
Comment (carmocca) [reply pytorch_lightning/trainer/training_loop.py]: Good point. I think that's okay

**Review by carmocca — COMMENTED** [review]
Pushed a few minor changes. Almost ready

**Comment by carmocca** [comment]
Changed the milestone to 1.4 as this requires several refactors not in the bug-fix branch

**Comment by simran2905** [comment]
Thanks @awaelchli and @carmocca for the merges.

**Review by carmocca — APPROVED** [review]

File: pytorch_lightning/trainer/connectors/optimizer_connector.py [hunk 14]
```diff
@@ -29,11 +29,17 @@ def on_trainer_init(self) -> None:
         self.trainer.optimizers = []
         self.trainer.optimizer_frequencies = []
 
-    def update_learning_rates(self, interval: str, opt_indices: Optional[List[int]] = None) -> None:
+    def update_learning_rates(
+        self, interval: str, update_plateau_schedulers: bool, opt_indices: Optional[List[int]] = None
```
Comment (Borda) [inline pytorch_lightning/trainer/connectors/optimizer_connector.py:33]: lest move the `update_plateau_schedulers` to the end ad make default value which kind of actual behaviour

File: pytorch_lightning/trainer/connectors/optimizer_connector.py [hunk 15]
```diff
@@ -46,6 +52,9 @@ def update_learning_rates(self, interval: str, opt_indices: Optional[List[int]]
             if isinstance(lr_scheduler['opt_idx'], int) and lr_scheduler['opt_idx'] not in opt_indices:
                 continue
 
+            if update_plateau_schedulers ^ lr_scheduler["reduce_on_plateau"]:
```
Comment (Borda) [inline pytorch_lightning/trainer/connectors/optimizer_connector.py:55]: just curious, what is this `^`? :)

File: tests/checkpointing/test_model_checkpoint.py [hunk 16]
```diff
@@ -262,6 +261,12 @@ def _make_assertions(epoch, ix, version=''):
         global_ix = ix + per_epoch_val_checks * epoch
         duplicated = bool(version)
 
+        # checkpoint saved at the end of training epoch will have updated lr_scheduler states
+        if epoch_aligned:
+            epoch_end_checkpoint = ix == (per_epoch_val_checks - 1)
+        else:
+            epoch_end_checkpoint = duplicated
```
Comment (Borda) [inline tests/checkpointing/test_model_checkpoint.py:268]:
```suggestion
        epoch_end_checkpoint = duplicated
        if epoch_aligned:
            epoch_end_checkpoint = ix == (per_epoch_val_checks - 1)
```

File: pytorch_lightning/trainer/connectors/optimizer_connector.py (hunk 14 above)
Comment (carmocca) [reply pytorch_lightning/trainer/connectors/optimizer_connector.py:33]: I requested that is changed to required as we want people to knowingly choose which schedulers to update.

File: pytorch_lightning/trainer/connectors/optimizer_connector.py (hunk 15 above)
Comment (carmocca) [reply pytorch_lightning/trainer/connectors/optimizer_connector.py:55]: XOR

**Review by justusschock — APPROVED** [review]

File: pytorch_lightning/loops/training_epoch_loop.py [hunk 17]
```diff
@@ -351,15 +357,13 @@ def _prepare_outputs(
...
 
-    def update_lr_schedulers(self, interval: str) -> None:
+    def update_lr_schedulers(self, interval: str, update_plateau_schedulers: bool) -> None:
         """updates the lr schedulers based on the given interval"""
-        if interval == "step":
-            finished_accumulation = self.batch_loop._accumulated_batches_reached()
-            finished_epoch = self._num_training_batches_reached()
-            if not finished_accumulation and not finished_epoch:
-                return
+        if interval == "step" and self.batch_loop.should_accumulate():
```
Comment (tchaton) [inline pytorch_lightning/loops/training_epoch_loop.py:362]: Neat !

**Review by tchaton — APPROVED** [review]
LGTM !

**Comment by tchaton** [comment]
Hey @simran2905,

quick question: Why differentiate reduce_on_plateau schedulers to the rest. Can't we call update_learning_rates only after callback metrics has been populated ?

Best,
T.C

**Comment by carmocca** [comment]
> quick question: Why differentiate reduce_on_plateau schedulers from the rest. Can't we call `update_learning_rates` only after callback metrics has been populated for all schedulers ?

See the discussion in the linked issue: https://github.com/PyTorchLightning/pytorch-lightning/issues/7637#issuecomment-854283411
